src/mame/drivers/firebeat.cpp

```diff
@@ -865,368 +1141,409 @@ WRITE_LINE_MEMBER(firebeat_state::spu_ata_dmarq)
 		}
 	}
 }
 
-WRITE_LINE_MEMBER(firebeat_state::spu_ata_interrupt)
-{
-	m_audiocpu->set_input_line(INPUT_LINE_IRQ6, state);
-}
-
-/*****************************************************************************/
-
-void firebeat_state::machine_start()
-{
-	/* set conservative DRC options */
-	m_maincpu->ppcdrc_set_options(PPCDRC_COMPATIBLE_OPTIONS);
-
-	/* configure fast RAM regions for DRC */
-	m_maincpu->ppcdrc_add_fastram(0x00000000, 0x01ffffff, false, m_work_ram);
-}
-
-void firebeat_state::firebeat_map(address_map &map)
+TIMER_CALLBACK_MEMBER(firebeat_spu_state::spu_dma_callback)
 {
-	map(0x00000000, 0x01ffffff).ram().share("work_ram");
-	map(0x70000000, 0x70000fff).rw(FUNC(firebeat_state::midi_uart_r), FUNC(firebeat_state::midi_uart_w)).umask32(0xff000000);
-	map(0x70006000, 0x70006003).w(FUNC(firebeat_state::extend_board_irq_w));
-	map(0x70008000, 0x7000800f).r(FUNC(firebeat_state::keyboard_wheel_r));
-	map(0x7000a000, 0x7000a003).r(FUNC(firebeat_state::extend_board_irq_r));
-	map(0x74000000, 0x740003ff).noprw(); // SPU shared RAM
-	map(0x7d000200, 0x7d00021f).r(FUNC(firebeat_state::cabinet_r));
-	map(0x7d000340, 0x7d000347).r(FUNC(firebeat_state::sensor_r));
-	map(0x7d000400, 0x7d000401).rw("ymz", FUNC(ymz280b_device::read), FUNC(ymz280b_device::write));
-	map(0x7d000800, 0x7d000803).r(FUNC(firebeat_state::input_r));
-	map(0x7d400000, 0x7d5fffff).rw("flash_main", FUNC(fujitsu_29f016a_device::read), FUNC(fujitsu_29f016a_device::write));
-	map(0x7d800000, 0x7d9fffff).rw("flash_snd1", FUNC(fujitsu_29f016a_device::read), FUNC(fujitsu_29f016a_device::write));
-	map(0x7da00000, 0x7dbfffff).rw("flash_snd2", FUNC(fujitsu_29f016a_device::read), FUNC(fujitsu_29f016a_device::write));
-	map(0x7dc00000, 0x7dc0000f).rw(m_duart_com, FUNC(pc16552_device::read), FUNC(pc16552_device::write));
-	map(0x7e000000, 0x7e00003f).rw("rtc", FUNC(rtc65271_device::rtc_r), FUNC(rtc65271_device::rtc_w));
-	map(0x7e000100, 0x7e00013f).rw("rtc", FUNC(rtc65271_device::xram_r), FUNC(rtc65271_device::xram_w));
-	map(0x7e800000, 0x7e8000ff).rw(m_gcu[0], FUNC(k057714_device::read), FUNC(k057714_device::write));
-	map(0x7fe00000, 0x7fe0000f).rw(FUNC(firebeat_state::ata_command_r), FUNC(firebeat_state::ata_command_w));
-	map(0x7fe80000, 0x7fe8000f).rw(FUNC(firebeat_state::ata_control_r), FUNC(firebeat_state::ata_control_w));
-	map(0x7ff80000, 0x7fffffff).rom().region("user1", 0);       /* System BIOS */
-}
+	// These values were picked because they roughly match real in-game footage of
+	// when sounds start playing, and more importantly to free up CPU time for the
+	// main CPU during DMAs to avoid stuttering.
+	auto next_dma_timing = attotime::from_usec(75);
+	auto dma_end_target = m_spu_ata_dma + 128;
 
-void firebeat_state::firebeat_spu_map(address_map &map)
-{
-	firebeat_map(map);
-	map(0x74000000, 0x740003ff).rw(m_dpram, FUNC(cy7c131_device::right_r), FUNC(cy7c131_device::right_w)); // SPU shared RAM
-}
+	while (m_spu_ata_dmarq && m_spu_ata_dma < dma_end_target)
```

Comment:
Responded on IRC but copy/pasting for transparency: That was more of an attempt to balance large DMAs and CPU time with the main CPU without causing hitching. It wants to transfer 0x1000000 to 0x2000000 bytes in one go and expects it to be done async in the background, which I was having a really hard time trying to make work in MAME.

I'll try to tweak the values a little more to see if using 512 bytes for the read size is doable without causing performance issues.

Reply:
I spent some time tweaking the values. Increasing the DMA size for each iteration has negative effects, even when adjusting the timing proportionally.

https://github.com/mamedev/mame/pull/7721/commits/213034b2fd2b5ca36c55b939fb77539c4a1933bf

I actually ended up going in the opposite direction. I imagine this will require a better PC to be able to handle properly, but smaller DMA blocks with a faster timer seems to offer better performance. I imagine because it's yielding more CPU time to the main CPU compared to before. Some of the smaller stuttering issues I was having seems to have cleared up with these new settings. These new settings are equivalent to reading about 240 bytes in the same 75 usec duration but it doesn't have the same stuttering issues that using those settings does.